docs/Eclipse4_Migration.md

```diff
@@ -0,0 +1,821 @@
+        // Replaces E3's enabledWhen expressions
+        return true;
+    }
+}
+```
+
+### Migration Steps
+
+1. **Define the handler in the model**:
+   - Open fragment.e4xmi in the E4 Model Editor
+   - Add a Model Fragment with Feature Name: `handlers`
+   - Set Parent Element ID: `org.eclipse.e4.legacy.ide.application` (or scope it to a specific part/window)
+   - Add a Handler element
+   - Set Contribution URI: `bundleclass://your.bundle.id/com.example.handlers.MyHandler`
+
+2. **Link to command**:
+   - If the command is in the same fragment, you can reference it directly
+   - If it's in a different fragment/application, create an import:
+     - Select "Imports" in the fragment
+     - Add a Command import
+     - Set the Element ID to your command's ID
+   - Set the handler's Command reference to your command
+
+3. **Convert handler class**:
+   - Remove `extends AbstractHandler` and `implements IHandler`
+   - Remove `execute(ExecutionEvent)` method signature
+   - Add `@Execute` annotation to execution method
+   - Use dependency injection instead of HandlerUtil:
+     ```java
+     // E3:
+     IWorkbenchWindow window = HandlerUtil.getActiveWorkbenchWindow(event);
+     ISelection selection = HandlerUtil.getCurrentSelection(event);
+     
+     // E4:
+     @Execute
+     public void execute(@Named(IServiceConstants.ACTIVE_SELECTION) ISelection selection,
+                        @Named(IServiceConstants.ACTIVE_SHELL) Shell shell) {
+         // Use injected values
+     }
+     ```
+
+4. **Convert enablement logic**:
+   - Replace `enabledWhen` expressions with `@CanExecute` method:
+     ```java
+     @CanExecute
+     public boolean canExecute(@Named(IServiceConstants.ACTIVE_SELECTION) ISelection selection) {
+         return selection != null && !selection.isEmpty();
+     }
+     ```
+
+5. **Handle active context** (replacement for `activeWhen`):
+   - In E4, handlers are scoped to their containing model element
+   - Global handlers: Add to application level
+   - Part-specific handlers: Add to the part's handler list
+   - Window-specific handlers: Add to the window's handler list
+
+6. **Remove E3 registration**: Remove the `<handler>` element from plugin.xml
+
+7. **Update MANIFEST.MF**: Ensure you import the injection packages:
+   ```
+   Import-Package: javax.inject;version="1.0.0",
+    javax.annotation;version="1.1.0"
+   ```
+
+### Common Injection Patterns for Handlers
+
+```java
+// Active shell
+@Named(IServiceConstants.ACTIVE_SHELL) Shell shell
+
+// Active part
+@Named(IServiceConstants.ACTIVE_PART) MPart part
+
+// Active selection
+@Named(IServiceConstants.ACTIVE_SELECTION) ISelection selection
+
+// Eclipse context
+IEclipseContext context
+
+// Part service
+EPartService partService
+
+// Model service
+EModelService modelService
+
+// Command service (for parameters)
+@Execute
+public void execute(@Named("parameter.name") String paramValue) {
+    // paramValue contains the command parameter
+}
+```
+
```

Comment:
It is important to mention here that handlers should always inject everything in the execute / canExecute method and not using fields injection to always get most recent injection values.